### ISSUE TITLE
Downloads vanish by themselves.

### ISSUE THREAD

**Issue by navid-zamani** [ram-on/SkyTube#886]
I had all then videos from this series downloaded and watched. But coming back a week later, half of them are gone.

https://youtu.be/xodtfM1r9FA

Only SkyTube could have deleted them. I did not delete them. They were not bookmarked either. (I only use bookmarks to mark videos for download later when on wifi.)

Sadly, some other important videos I wanted to keep, are gone too, I and don't know how to find them again. Which is really bad, as they were important for my life.

Can you remove this... what seems to be an automatic "feature"?
Thanks.

(Workaround for now: I use Ghost Commander, to move them to a different directory.)

**Comment by gzsombor** [comment]
SkyTube only deletes file, when you explicitly click on the 'Delete downloaded file' menu. And if you enabled the separate folders for different channels, and that folder doesn't have any file in it, the app will delete the folder too.
You can move your files from the Downloads/Movies directory, and SkyTube will detect, that the file disappeared, and remove the video from the Downloads tab.
  Are you sure, that you haven't run some automatic 'System Optimizer/Cleanup/Cloud Sync' software, which determined, that these files are large, and never used, so it's worthy for cleanup?

**Comment by navid-zamani** [comment]
Hey,

I thought so too, but I wouldn’t have created this ticket, unless I was reasonably certain I did not click delete.

I also generally don’t run anything on here that isn’t truly necessary. Let alone anything that I can’t fully trust or might do things behind my back.

To make absolutely sure, I will do a deliberate test with some videos, and report back when I have the results. (Including watching if they disappear even with SkyTube completely stopped.)

**Comment by gzsombor** [comment]
Yeah, totally understand your concerns, it's definitely looks strange, but AFAIK the places, where the app actually tries to delete anything from the filesystem:
* when you click on the delete menu item - then it deletes the video file
* when it detects, that the directory, where the downloaded video was sitting, is empty now, and the 'separate folder for channels' is on
* when it found out, that someone deleted the video/audio file, and it tries to delete the audio/video file. (Currently it's not a valid scenario, as the app doesn't support downloading separate files for audio and video content)
* after an update downloaded an APK file, it cleans up the old ones.

Based on this, I don't think that any of the code could go rogue and delete your files overnight. Maybe, if you uninstall the app, sometimes Android also deletes all the files, which were created by that app - but I dont know for sure, when this could happen.

**Comment by acanoe** [comment]
It also happens to me, I thought I toggled the wrong option.
The video files I downloaded are all gone, leaving only the folders (I turned on separate folders for channels option).
From what I've experienced, the videos are auto-deleted after one/two weeks.
I'm using the SkyTube Extra app and I set it to save the videos to external SD.

**Comment by gzsombor** [comment]
Staring on the code longer, I could only discover smaller problems - cases where the app deletes an empty folder, or when it doesn't delete an empty folder, but it should.

To try to reproduce it in emulator, could you please share your Android version? The path which you set as a download folder? If not, is it /storage/emulated/0/Movies ? Or something else? Have you set the 'Download Videos Into Folders' checkbox? Have you changed any of these settings after you downloaded these videos and their accidental disappearance?

**Comment by acanoe** [comment]
In my case, it's Android 10 with Samsung's OneUI v2.0, I set the download folder to /storage/2302-1E1A/Movies (external SD). I also checked "Download videos into folders" option before downloading any videos, but the videos got deleted leaving only their folders, I then set the option to off, tried to download other videos, but they still got deleted. There are other mp4 files I downloaded from somewhere else but weirdly they weren't deleted, only those downloaded from the app.

**Comment by acanoe** [comment]
Could it be because of this? 
https://stackoverflow.com/questions/46501822/downloaded-files-get-deleted-automatically

**Comment by gzsombor** [comment]
I tried to reproduce it in the emulator - and I couldn't. However, I see problems with not able to delete the files by the app - but these errors are silently ignored by the app currently.

**Comment by gzsombor** [comment]
Hmm, that's interesting, yes, that could happen too.
 It seems, that newer Androids, when a file was downloaded by the built-in DownloadManager, it is considered to be 'owned' by that service, and SkyTube couldn't delete it. Neither the folders.

**Comment by acanoe** [comment]
Maybe try downloading to internal data directory then moving the files to the final download path, renaming the files i? Will that possibly change the "ownership" of the files?

**Comment by acanoe** [comment]
Maybe try downloading the files to the app's internal data dir first, and then after that, moving (and renaming) the files to the final download directory? Will that change the file "ownership"?

**Comment by gzsombor** [comment]
... and I could trigger the error with setting the date to be in the future, 
and execute `cmd jobscheduler run -f com.android.providers.downloads -100`

The only strange thing is that I have device with Android 10, but I haven't seen this disappearing downloads error.

**Comment by navid-zamani** [comment]
@gzsombor:

Hey,

I just had the opposite error: I downloaded a video with SkyTube, and while it is in he file system and I can watch it with VLC, SkyTube doesn't list it under Downloads. Which makes me think the download manager does not behave as it should and as your app expects it.

And sorry for not providing details on my system with the bug. Shameful, with me being an old programmer myself. :/
My OS is Huawei EMUI build 10.0.0.182, which is also Android 10.
I run SkyTube 2.976 from 2021-01-08 06:28.
EMUI is sometimes a bit rough around the edges.
I use the default /storage/emulated/0/Movies directory. With separate folders kept off. And since the bug report, I only changed the resolution settings. (Thanks for the update by the way. Saldy, it misses the problem I had with it, due to not muxing audio/video itself, but it is appreciated.)

Another thing I noticed that may cause the download manager to behave weirdly, is that sometimes, it does not recognize if I just connected to pr discommected to wifi. Especially if two connect/disconnect events happen rapidly, like wifi being lost and 4G+ taking over.

**Comment by navid-zamani** [comment]
I also have trouble reproducing this.

I would not put any more work into it, until somebody can offer something reproducible. I will try to do that, but right now, none of the three files I downloaded just to see if they would vanish actually vanished. I will download some more in the next days, while playing around with connection settings and times and throttled connections.
(My current situation is usually a slow 384kB/s shared wifi that somelimes loses reception and only is available during work hours, and a throttled 4G+ connection at 8kB/s for the rest. If that helps.)

**Comment by navid-zamani** [comment]
Another thing I noticed, is that you cannot rely on the download manager to remember everything you wanted to download. Nor rely on it not remembering.
When I was new to SkyTube, months ago, I just used "download" like bookmarking, thinking it would eventually download them even if the connection got lost. But often, it just didn't.
But in a recent case, I assumed it would not remember, even after 2 minutes, yet it did, and as a result, multiple taps to "download" over 15 minutes resulted in multiple downloads even though I thought I had triggered a cancellation and forgetting.
In both cases, very likely not the fault of your app but of the download manager. But trying to get a Chinese corporation to listen to your reports and fix your bugs is a unrealistic dream. ;)

I guess there is a reason, Firefox implements its own internal download manager. (Which you would need for muxing anyway.)

Have you ever looked into the Firefox extension "Video Download Helper?" It is very good at downloading even weird chunked YT stream formats. I think the code is open, so learning from it should be easy. It's what open source is for, after all. :)
(I also does a LOT of other video sites.)

**Comment by acanoe** [comment]
> The only strange thing is that I have device with Android 10, but I haven't seen this disappearing downloads error.

Does the method of which the app is installed matter? E.g. via adb vs sideloading from the file manager/browser/f-droid app.

**Comment by acanoe** [comment]
> In both cases, very likely not the fault of your app but of the download manager. But trying to get a Chinese corporation to listen to your reports and fix your bugs is a unrealistic dream. ;)

I think you have a point here, maybe some Android ROMs are modifying how the DownloadManager works. In that StackOverflow question somebody wrote this:

> There is a bigger problem starting in Android Q where value set by `setVisibleInDownloadsUi` is ignored and it is false by default.

Maybe on some ROMs, the value is not set, defaulting to false, and some others are set to true? Hence in some Android 10 phones the files are deleted and in some others they aren't?

**Comment by navid-zamani** [comment]
One way around it, would be to put downloads under Download/SkyTube, as Download is exempt from auto-deletion.

Another thing I found that may play a role, is that on Huawei devices and some other manifacturers’, apps ar automatically closed and blocked from launching in the background by default, to save power. So if your app needs to run in any way to successfully finish the entire download process, you jeed to tell users to set it to be manually managed and allow that type of running.

I must say I have not yet seen another new case of auto-deletion yet. But I don’t trust it anymore  ;)

**Comment by gzsombor** [comment]
Yes, unfortunately I've already saw a couple of problems with Android's builtin DownloadManager:
* As SkyTube doesn't save the download requests persistently, if you start a couple of downloads, switch to other app, and the app is closed in the background by the OS, when you open up SkyTube again, you won't see the newly downloaded videos, even if the download was successfully finished.
* If you select the external SD card as the download folder, in newer androids, the DownloadManager is unable to write to there.  

And it seems, that there is - at least - 2 other issue, which we've discovered in this thread, nice :smile:

**Comment by navid-zamani** [comment]
I think can solve the second point:
On those phones, no app can write to the external SD with the normal interface. But if you use the storage access framework, they can. At least with my file manager, Ghost Commander, that was how I could write to it anyway.

**Comment by gzsombor** [comment]
Yes, that could be the case!
Currently, the app has the option to specify the exact download directory, which makes it hard to make the download process bug-free.
If the user specifies:
* Environment.getExternalStoragePublicDirectory(DIRECTORY_DOWNLOADS) - no hacks are needed
* Environment.getExternalStoragePublicDirectory(DIRECTORY_MOVIES) - we would need to rename the files to hide it from the reaper - DownloadManager's cleanup process
* External SD card's Movies or Downloads folder - we need to use that Storage Access Framework to read/write/delete files and directories
* else ? - unknown.

Currently I don't know any reason, why anyone would want to specify some other custom locations, so maybe, one solution could be to limit the download destinations to these cases, so these hacks can be implemented specifically for each case. What do you think?

**Comment by acanoe** [comment]
I think putting all downloads into the Movies folder and giving the user a choice between saving the videos into internal storage Movies folder or external is enough, they can still view all their downloaded videos in the app anyway.

**Comment by navid-zamani** [comment]
I'd just always use Downloads as a temporary directory for the download manager, and move them to the right directory, always using SAF, during the takeover from "download manager is finished" to "SkyTube reacts to notification of it being finished".
That way, no special code would be needed and all direcories could be handled by the same code.

**Comment by navid-zamani** [comment]
@acanoe: I found rewind control in SkyTube to be very uncomfortable for my use case, as it always rewinds too little or way too much. Because double tapping to go back a confgurable amount (10 or 5 seconds) is missing.

I also see no need to duplicate an entire good video player when I already have one on my system (VLC). That only exists because YouTube has a weird streaming system (using chunks) that stops one from simply handing the URL to another player. But if possible, like with downloads, I would like to always use my own player.

Also, due to this download manager hassle, I've got downloads that do not show in SkyTube, and other downloads that still show, despite me deleting them already with other apps. So access from other apps should not be prevented.

**Comment by navid-zamani** [comment]
Okay, the three downloads I kept as canaries just vanished in the last days.

I think it was during the **update**! (I'm using F-Droid.)

The next time I've got proper wifi, I will try downgrading, downloading something, and then upgrading, and report if I can reliably reproduce this.

**Comment by gzsombor** [comment]
Just for reference, these are the already known bugs: #529 and #681

**Comment by maszek** [comment]
Hello Team,
Has there been an update on this bug? Because I am still experiencing this behavior. It also occurs with some bookmarks. Some random videos I have bookmarked simply vanish from my list of bookmarks.

**Comment by navid-zamani** [comment]
@maszek: As a workaround, as I said, set the directory to something inside of your Android's ”Download(s)“ directory, as that is protected from Android's cleanup routine.

And yes, this should be fixed, because AFAIK it is an easy fix, by just telling Android to exempt the default directory in the manifest, or by setting the directory like above by default.

**Comment by hamishmb** [comment]
I experienced this at the end of June 2022, where all my downloads disappeared. I checked through all my folders as well and they were actually gone.

Android 11 here on a Moto G8 Power, which I think means it's the download manager bug causing this problem for me? Shall I try to reproduce by leaving something downloaded for a while?

Any news on this issue?

**Comment by navid-zamani** [comment]
> Any news on this issue?

Please read the comments here. The problem is that you and I were dumb enough to buy an Android phone. Even though we knew it acts an an abusive condescending nanny and is made by the criminal organization “Google” (crimes: mass mental abuse (“advertisement“) for the purpose of robbery/theft (“profit“)). And one of these condescending properties is, that it does not let you decide what your own damn device (formerly a computer) does and what not. And one of those things it pushed down its victims throats is that it outright “deletes” *everything* that isn’t in some manifest/whitelist of an app, with Downloads being the only exception. (There is actually no such function as deletion. It only knows reading and writing. What it actually does, is overwrite the pointers to the files, that are stored in the directories and in an allocation table of some sorts. That is why on a real computer, you can still just instantly remove power to prevent any further writes, boot the system with an external OS with only read access to that file system, and use a computer forensics tool to still read all the files from it, provided you have some kind of signature/header that you can recognize the file data by. (E.g. “JFIF” for JPEGs). So nothing is actually gone until you overwrite it. You just have to find it again.)

**So that is the solution here: Configure SkyTube to store all its downloads in Downloads/SkyTube/ for example. Works well for me.**

_[Yeah, today I’m not in the mood to appease complete pieces of shit nor find excuses for why we’re so fucking stupid to play along and become domesticated livestock. (Nor to appease the child rapists aka religious that want to ban us from swearing as a result of their schizophrenic delusions and projecting their self-hate onto us.) ^^]_
